File: mawhrin/items/book/book.go
// Package book implements the Book type and methods.
package book

import (
	"os"

	"github.com/stvmln86/mawhrin/mawhrin/items/note"
	"github.com/stvmln86/mawhrin/mawhrin/tools/file"
	"github.com/stvmln86/mawhrin/mawhrin/tools/neat"
	"github.com/stvmln86/mawhrin/mawhrin/tools/path"
)

// Book is a single directory containing multiple Notes.
type Book struct {
	Dire string
	Extn string
	Mode os.FileMode
}

// New returns a new Book.
func New(dire, extn string, mode os.FileMode) *Book {
	dire = neat.Path(dire)
	extn = neat.Extn(extn)
	return &Book{dire, extn, mode}
}

// Create returns a new Note containing a string.
func (b *Book) Create(name, body string) (*note.Note, error) {
	name = neat.Name(name)
	body = neat.Body(body)
	dest := path.Join(b.Dire, name, b.Extn)
	if err := file.Create(dest, body, b.Mode); err != nil {
		return nil, err
	}

	return note.New(dest, b.Mode), nil
}

// CreateOrGet returns a new (and empty) or existing Note.
func (b *Book) CreateOrGet(name string) (*note.Note, error) {
	name = neat.Name(name)
	dest := path.Join(b.Dire, name, b.Extn)
	if !file.Exists(dest) {
		if err := file.Create(dest, "", b.Mode); err != nil {
			return nil, err
		}
	}

	return b.Get(name)
}

// Filter returns all existing Notes passing a filter function.
func (b *Book) Filter(ffun func(*note.Note) (bool, error)) ([]*note.Note, error) {
	notes, err := b.List()
	if err != nil {
		return nil, err
	}

	var goods []*note.Note
	for _, note := range notes {
		ok, err := ffun(note)
		switch {
		case err != nil:
			return nil, err
		case ok:
			goods = append(goods, note)
		}
	}

	return goods, nil
}

// Get returns an existing Note.
func (b *Book) Get(name string) (*note.Note, error) {
	name = neat.Name(name)
	orig := path.Join(b.Dire, name, b.Extn)
	orig, err := file.Get(orig)
	if err != nil {
		return nil, err
	}

	return note.New(orig, b.Mode), nil
}

// List returns a sorted slice of all existing Notes.
func (b *Book) List() ([]*note.Note, error) {
	origs, err := path.Glob(b.Dire, b.Extn)
	if err != nil {
		return nil, err
	}

	var notes []*note.Note
	for _, orig := range origs {
		notes = append(notes, note.New(orig, b.Mode))
	}

	return notes, nil
}

// Match returns all existing Notes with names containing a substring.
func (b *Book) Match(term string) ([]*note.Note, error) {
	return b.Filter(func(note *note.Note) (bool, error) {
		return note.Match(term), nil
	})
}

// Search returns all existing Notes with bodies containing a substring.
func (b *Book) Search(term string) ([]*note.Note, error) {
	return b.Filter(func(note *note.Note) (bool, error) {
		return note.Search(term)
	})
}

File: mawhrin/items/book.py
"""
Class definition for 'Book'.
"""

from collections.abc import Callable, Iterator

from mawhrin import tools
from mawhrin.items.note import Note


class Book:
    """
    A directory containing multiple plaintext note files.
    """

    def __init__(self, dire: str, extn: str):
        """
        Initialise the Book.
        """

        self.dire = tools.data.path(dire)
        self.extn = tools.data.extn(extn)

    def __contains__(self, name: str) -> bool:
        """
        Return True if the Book contains a named Note.
        """

        return self.get(name) != None

    def __eq__(self, book: object) -> bool:
        """
        Return True if the Book is equal to another Book.
        """

        if not isinstance(book, Book):
            return NotImplemented
        else:
            return self.dire == book.dire and self.extn == book.extn

    def __iter__(self) -> Iterator[Note]:
        """
        Yield each Note in the Book in alphabetical order.
        """

        for path in tools.path.glob(self.dire, self.extn):
            yield Note(path)

    def __len__(self) -> int:
        """
        Return the number of Notes in the Book.
        """

        iter = tools.path.glob(self.dire, self.extn)
        return len(list(iter))

    def __repr__(self) -> str:
        """
        Return the Book as a code-representative string.
        """

        return f"Book({self.dire!r}, {self.extn!r})"

    def create(self, name: str) -> Note:
        """
        Return a new or existing Note in the Book.
        """

        name = tools.data.name(name)
        path = tools.path.join(self.dire, name, self.extn)
        return Note(path)

    def exists(self) -> bool:
        """
        Return True if the Book's directory exists.
        """

        return tools.file.exists(self.dire, dire=True)

    def get(self, name: str) -> Note | None:
        """
        Return an existing Note in the Book, or None.
        """

        note = self.create(name)
        return note if note.exists() else None
